CMake: unbreak ENABLE_SYSTEM_GMIC

amyspark opened the pull request:
This reverts upstream commit https://github.com/c-koi/gmic-qt/commit/f0d9d8acd10b89f9d28a49227f3c70ff1f2e18a5 .

[BUG:466070](https://bugs.kde.org/466070)

@darix can you confirm this fixes packaging on your end?

darix commented:
i asked in the other commit also why it shouldnt be possible anymore.

darix commented:
JFYI: i just did a test build on our side... and our gmic package still properly builds all the plugins and properly links it against the shared library. though it should be mentioned we carry your changes on top of gmic as krita5.patch and then build the krita plugin as part of the gmic build. so that's why the openSUSE build doesnt notice this probably.

darix commented:
https://discuss.pixls.us/t/building-against-system-gmic-broken-since/35438?u=darix

antonio-rojas commented:
With this patch, build fails with a different error
```
/build/krita-plugin-gmic/src/gmic-3.2.1.1-patched/gmic-qt/src/PersistentMemory.cpp: In static member function ‘static void GmicQt::PersistentMemory::clear()’:
/build/krita-plugin-gmic/src/gmic-3.2.1.1-patched/gmic-qt/src/PersistentMemory.cpp:43:17: error: no matching function for call to ‘gmic_library::gmic_image<char>::assign()’
   43 |   image().assign();
      |   ~~~~~~~~~~~~~~^~
In file included from /build/krita-plugin-gmic/src/gmic-3.2.1.1-patched/gmic-qt/src/PersistentMemory.cpp:26:
/usr/include/gmic.h:111:20: note: candidate: ‘template<class t> gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::assign(const t*, unsigned int, unsigned int, unsigned int, unsigned int) [with T = char]’
  111 |     gmic_image<T>& assign(const t *const values, const unsigned int size_x, const unsigned int size_y=1,
      |                    ^~~~~~
/usr/include/gmic.h:111:20: note:   template argument deduction/substitution failed:
/build/krita-plugin-gmic/src/gmic-3.2.1.1-patched/gmic-qt/src/PersistentMemory.cpp:43:17: note:   candidate expects 5 arguments, 0 provided
   43 |   image().assign();
      |   ~~~~~~~~~~~~~~^~
/usr/include/gmic.h:106:20: note: candidate: ‘gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::assign(unsigned int, unsigned int, unsigned int, unsigned int) [with T = char]’
  106 |     gmic_image<T>& assign(const unsigned int size_x, const unsigned int size_y=1,
      |                    ^~~~~~
/usr/include/gmic.h:106:20: note:   candidate expects 4 arguments, 0 provided
/usr/include/gmic.h:114:20: note: candidate: ‘gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::assign(const T*, unsigned int, unsigned int, unsigned int, unsigned int, bool) [with T = char]’
  114 |     gmic_image<T>& assign(const T *const values, const unsigned int size_x, const unsigned int size_y,
      |                    ^~~~~~
/usr/include/gmic.h:114:20: note:   candidate expects 6 arguments, 0 provided
```

amyspark commented:
It looks like the particular iteration of `gmic_image::assign()` G'MIC uses requires access to the private API of libgmic, which is now disabled when `ENABLE_SYSTEM_GMIC` is on.

amyspark commented:
@antonio-rojas The commit now injects the private extension into the library (through a raw `include_directories`, because I'm not sure what kind of library the imported `gmic` target is). Let me know if that suffices.

amyspark committed:
CMake: actually fix ENABLE_SYSTEM_GMIC

PR c-koi/gmic-qt#172 is correct in that it is no longer possible
to build without -Dgmic_core, due to G'MIC-Qt relying in
now-private implementation details of G'MIC; namely, the extensions
made in gmic.cpp to CImg.

However, the solution chosen has significant shortcomings:

1. It blindly assumes that the consumed library has been built by a
  GCC-compatible compiler. This is easily inferred from the lack of
  symbol exports in {CImg,gmic}.{h,cpp}.
2. It makes no provision for the exported library type; G'MIC can be
  built statically or dynamically.
3. In Windows, when built with MSVC, the kind of symbol export that
  gmic_core implies is only available with a static libgmic.

To fix this, this commit augments G'MIC-Qt's `ENABLE_DYNAMIC_LINKING`
handling with target detection code for the above described cases.
In the case where a compatible library is not found, a fallback is
specified that will build libgmic as a separate target, then make
G'MIC-Qt link against it in order to mimic the requirements. If a
suitable system library is found, we augment it with the gmic.cpp
plugin to make the symbols visible at compile time.

darix commented:
tbh the real solution is that David moves the gmic_image stuff into a public header as it should be. maybe we could prepare a submission for this as a PR to gmic?

antonio-rojas commented:
Getting a different issue now
```
CMake Error at /usr/lib/cmake/gmic/GmicTargets.cmake:70 (set_target_properties):
  The link interface of target "libgmic" contains:

    CImg::CImg

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /usr/lib/cmake/gmic/GmicConfig.cmake:36 (include)
  CMakeLists.txt:145 (find_package)
```

Given that upstream decided to completely stop supporting cmake, I'm afraid the only option to use system gmic would be to ship a custom findGmic.cmake. For now I'm just going to patch downstream to manually link to system gmic.

antonio-rojas commented:
Managed to "fix" it by making the CImg::CImg link target private in gmic's cmake config. After that, the krita plugin builds fine.

amyspark commented:
I've removed the usage of the official (broken) targets and simplified the fallback mechanism. If you set `-DENABLE_DYNAMIC_LINKING=ON -DENABLE_SYSTEM_GMIC=OFF` you'll link against the emulated libgmic. Set `-DENABLE_SYSTEM_GMIC=ON` to preserve the existing behaviour and inject the plugin into the headers.
